python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_ledger_orchestrator/_ledger_orchestrator_manager.py

```diff
@@ -0,0 +1,333 @@
+
+    async def select_speaker(self, thread: List[AgentMessage]) -> str:
+        """Select a speaker from the participants and return the
+        topic type of the selected speaker."""
+        return ""
+
+    async def reset(self) -> None:
+        """Reset the group chat manager."""
+        pass
+
+    async def on_unhandled_message(self, message: Any, ctx: MessageContext) -> None:
+        raise ValueError(f"Unhandled message in group chat manager: {type(message)}")
+
+    async def _reenter_inner_loop(self):
+        # Reset the agents
+        await self.publish_message(
+            GroupChatReset(),
+            topic_id=DefaultTopicId(type=self._group_topic_type),
+        )
+        self._message_thread.clear()
+
+        # Prepare the ledger
+        ledger_message = TextMessage(
+            content=self._get_task_ledger_full_prompt(self._task, self._team_description, self._facts, self._plan),
+            source=self._name,
+        )
+
+        # Save my copy
+        self._message_thread.append(ledger_message)
+
+        # Log it
+        await self.publish_message(
+            GroupChatMessage(message=ledger_message),
+            topic_id=DefaultTopicId(type=self._output_topic_type),
+        )
+
+        # Broadcast
+        await self.publish_message(
+            GroupChatAgentResponse(agent_response=Response(chat_message=ledger_message)),
+            topic_id=DefaultTopicId(type=self._group_topic_type),
+        )
+
+        # Restart the inner loop
+        await self._orchestrate_step()
+
+    async def _orchestrate_step(self) -> None:
+        # Check if we reached the maximum number of rounds
+        if self._n_rounds > self._max_rounds:
+            await self._prepare_final_answer("Max rounds reached.")
+            return
+        self._n_rounds += 1
+
+        # Update the progress ledger
+        context = self._thread_to_context()
+
+        progress_ledger_prompt = self._get_progress_ledger_prompt(
+            self._task, self._team_description, self._participant_topic_types
+        )
+        context.append(UserMessage(content=progress_ledger_prompt, source=self._name))
+
+        response = await self._model_client.create(context, json_output=True)
+
+        assert isinstance(response.content, str)
+        progress_ledger = json.loads(response.content)
+
+        # Check for task completion
+        if progress_ledger["is_request_satisfied"]["answer"]:
+            await self._prepare_final_answer(progress_ledger["is_request_satisfied"]["reason"])
+            return
+
+        # Check for stalling
+        if not progress_ledger["is_progress_being_made"]["answer"]:
+            self._n_stalls += 1
+        elif progress_ledger["is_in_loop"]["answer"]:
+            self._n_stalls += 1
+        else:
+            self._n_stalls = max(0, self._n_stalls - 1)
+
+        # Too much stalling
+        if self._n_stalls >= self._max_stalls:
+            await self._update_task_ledger()
```

Comment:
should we add a logging event that there is a replan?

Reply:
I think we need to a add a lot of logging. I'm just not sure how to do that. We've got several channels I guess.... group_topic, output_topic, and at least two loggers

Reply:
I think this should be new events provided to the output topic.